src/dbnode/persist/fs/files.go

```diff
@@ -373,51 +415,136 @@ func (a fileSetFilesByTimeAndVolumeIndexAscending) Less(i, j int) bool {
+				break
+			}
+		}
 	}
-	str := strings.Replace(components[1], fileSuffix, "", 1)
-	nanoseconds, err := strconv.ParseInt(str, 10, 64)
+
+	return delimPos, delimsFound
+}
+
+// Returns the the specified component of a filename, given the positions of
+// delimeters. Our only use cases for this involve extracting numeric
+// components, so this function assumes this and returns the component as an
+// int64.
+func intComponentFromFilename(
+	baseFilename string,
+	componentPos int,
+	delimPos [maxDelimNum]int,
+) (int64, error) {
+	var start int
+	if componentPos > 0 {
+		start = delimPos[componentPos-1] + 1
+	}
+	end := delimPos[componentPos]
+	if start > end || end > len(baseFilename)-1 || start < 0 {
+		return 0, fmt.Errorf(errUnexpectedFilename, baseFilename)
+	}
+
+	num, err := strconv.ParseInt(baseFilename[start:end], 10, 64)
 	if err != nil {
-		return nil, timeZero, err
+		return 0, fmt.Errorf(errUnexpectedFilename, baseFilename)
 	}
-	return components, time.Unix(0, nanoseconds), nil
+	return num, nil
 }
 
 // TimeFromFileName extracts the block start time from file name.
 func TimeFromFileName(fname string) (time.Time, error) {
-	_, t, err := componentsAndTimeFromFileName(fname)
-	return t, err
+	base := filepath.Base(fname)
+
+	delims, delimsFound := delimiterPositions(base)
+	if delimsFound < 2 {
+		return timeZero, fmt.Errorf(errUnexpectedFilename, fname)
+	}
+	nanos, err := intComponentFromFilename(base, timeComponentPosition, delims)
+	if err != nil {
+		return timeZero, fmt.Errorf(errUnexpectedFilename, fname)
+	}
+
+	return time.Unix(0, nanos), nil
 }
 
-// TimeAndIndexFromCommitlogFilename extracts the block start and index from file name for a commitlog.
+// TimeAndIndexFromCommitlogFilename extracts the block start and index from
+// file name for a commitlog.
 func TimeAndIndexFromCommitlogFilename(fname string) (time.Time, int, error) {
 	return timeAndIndexFromFileName(fname, commitLogComponentPosition)
 }
 
-// TimeAndVolumeIndexFromFileSetFilename extracts the block start and volume index from file name.
+// TimeAndVolumeIndexFromDataFileSetFilename extracts the block start and volume
+// index from a data fileset file name that may or may not have an index. If the
+// file name does not include an index, unindexedFilesetIndex is returned as the
+// volume index.
+func TimeAndVolumeIndexFromDataFileSetFilename(fname string) (time.Time, int, error) {
+	base := filepath.Base(fname)
+
+	delims, delimsFound := delimiterPositions(base)
+	if delimsFound < 3 {
+		return timeZero, 0, fmt.Errorf(errUnexpectedFilename, fname)
+	}
+
+	nanos, err := intComponentFromFilename(base, timeComponentPosition, delims)
+	if err != nil {
+		return timeZero, 0, fmt.Errorf(errUnexpectedFilename, fname)
+	}
+	unixNanos := time.Unix(0, nanos)
+
+	// Legacy filename with no volume index.
+	if delimsFound == 3 {
+		return unixNanos, unindexedFilesetIndex, nil
+	}
+
+	volume, err := intComponentFromFilename(base, 2, delims)
```

Comment:
should we make these positions constants? I could go either way.

Reply:
Yeah makes sense, since the other places use position constants too.